### PR TITLE
Add a 'domain' field to validator lists:

### PR DESCRIPTION
**Pull request by cryptobrad**
This patch adds a new optional field to validator lists allowing
a validation list publisher to include the domain associated with
a validator it is listing.

I also rewrote an awkward loop using C++ constructs.

### CODE REVIEW

**Comment by MarkusTeufelberger** [comment]
I disagree with the concept by the way - owning a domain in the DNS system and running a validator are not linked in any way nor should they be. I also don't want some random (potentially malicious/illegal) domains showing up on my machine, especially since this is essentially a completely free text field.

Lastly, there is no verifiable proof attached to domain ownership in this format.

**Comment by cryptobrad** [comment]
A team publishing validator list can choose to show the domain of the validator in the list. If the group lies there is proofs they lied! The validator list it is signed by them. That is good! Now there is no evidence and we must go to xrpcharts and search for the identity in the list which can be changed.

If you don’t want the random (potentially malicious/illegal) domains showing up on the machine then you maybe don’t want to be running rippled or bitcoind anyways. Ripple root accout has a domain and all the transactions can contain arbitrary text.

Is it better if I use another name, like Arnaud Nonymous or Jemes Bond 0-0-7 and photoshop that head on the Mr. Robot body? I choose to be anonymous for reasons. If the Ripple team don't want my code because of my picture and funny name thats fine. I can contribute to other code.

**Comment by nbougalis** [comment]
Thanks for your submission. We'll run this through a review and CI.

**Comment by HowardHinnant** [comment]
I'm getting a repeatable unittest failure in `ValidatorRPC`:
```
#22 failed: unhandled exception: Type is not convertible to string
```

**Comment by cryptobrad** [comment]
I forgot the change of the unit tests in my commit sorry. It is fixed now.

File: src/ripple/app/misc/impl/ValidatorList.cpp [hunk 1]
```diff
@@ -176,7 +176,10 @@ ValidatorList::load (
         // Config listed keys never expire
         if (it.second)
             it.first->second.expiration = TimeKeeper::time_point::max();
-        it.first->second.list.emplace_back(std::move(*id));
+        if (match.size() == 3)
+            it.first->second.list.emplace_back(*id, match[2]);
+        else
+            it.first->second.list.push_back(*id);
```
Comment (HowardHinnant) [inline src/ripple/app/misc/impl/ValidatorList.cpp]: I don't think this code is doing what you think it is doing.  `match.size()` is a function only of the regex, and not of the search string.  And the regex (`re`) is `const`.  `match.size()` is _always_ 3.

File: src/ripple/app/misc/impl/ValidatorList.cpp [hunk 2]
```diff
@@ -202,80 +205,90 @@ ValidatorList::applyList (
...
+
+                publisherList.emplace_back(
+                    PublicKey{Slice{ret.first.data(), ret.first.size()}},
+                    domain);
             }
 
-            if (val.isMember ("manifest") && val["manifest"].isString ())
-                manifests.push_back(val["manifest"].asString ());
+            if (val.isMember("manifest") && val["manifest"].isString())
+                manifests.push_back(val["manifest"].asString());
```
Comment (HowardHinnant) [inline src/ripple/app/misc/impl/ValidatorList.cpp]: Is it an intentional change that this code now gets executed when `val` doesn't have a "validation_public_key", but before this code did not?

File: src/ripple/app/misc/impl/ValidatorList.cpp [hunk 3]
```diff
@@ -202,80 +205,90 @@ ValidatorList::applyList (
...
+
+        // Remove all entries that are not longer referenced
+        for (auto it = keyListings_.begin(); it != keyListings_.end();)
         {
-            ++iNew;
-            ++iOld;
+            if (it->second <= 0)
+                it = keyListings_.erase(it);
+            else
+                ++it;
```
Comment (HowardHinnant) [inline src/ripple/app/misc/impl/ValidatorList.cpp]: While the new code is more readable, and I like the use of the std::algorithms, this turns a 1-pass algorithm into a 3-pass algorithm.  I'm not sure it is worth the performance hit.

Comment (nbougalis) [reply src/ripple/app/misc/impl/ValidatorList.cpp]: I share your concern generally, but I don't think this code is invoked frequently enough to really matter. And the new code is, as you say, much more readable and (imo) less error prone. I had to read the existing code a couple of times to _really_ convince myself it does what it claims to.

**Comment by nbougalis** [comment]
>I disagree with the concept by the way - owning a domain in the DNS system and running a validator are not linked in any way nor should they be. 

Tell me, which of these three validators is operated by Banhof? No cheating and looking at charts...

- `nHUzum747yqip3HWSgzSNHNMjmLUqhroNVWidSRTREswEVhKNQEM`? 
- `nHUUdvRVh4a3oimLveoHbuT99q2y425pV6QxHW6qyogeYzRpkcM8`? 
- `nHULGMbQHyXe92wtnxA1X9TEp26qPiRfwfTv5MskD1yMro7bQ2df`?

I agree that a domain shouldn't be a _prerequisite_ for operating a validator, and it **isn't**. But we need **some** kind of identifier in a common "regulated" namespace that we all share that allows us to assign and group validators by human-readable and human-relevant names. Right now, I can think of only two such identifiers:

1. a domain name, which is what we are using; or
2. an email address.

Between the two, I think the domain name is preferable for a number of reasons. It's possible there are better options, and  if you know of any then by all means... let's hear them.

>Lastly, there is no verifiable proof attached to domain ownership in this format.

Right now, given a published list, the only way to associate a validator public key with an identifier is by visiting a website; this is a pain. Additionally, the data that website reports is not signed in an way.

This change actually improves things: the publisher is attesting that a given validator corresponds to a particular domain and _cryptographically signs that statement_; if the publisher of the list lies, there will be cryptographic proof of that lie and we can all decide what to do moving forward.

It doesn't get us to the point of users being able to independently verify the domain by posting a signature or other cryptographic proof that binds the domain and the validator together, but that is something we _are_ working on. 

We are working on a way to allow people to independently verify the domain (by posting the signature) but in the mean time, I see this as a net win.

**Comment by MarkusTeufelberger** [comment]
> Tell me, which of these three validators is operated by Banhof? No cheating and looking at charts...

Easy: None of them is, because Ba**h**nhof is not operating a validator.

Proof: https://www.bahnhof.se/ripple.txt

Alternatively https://www.google.com/search?q=site%3Abahnhof.se+ripple also only finds https://www.bahnhof.se/press/press-releases/2017/07/18/fintechbolaget-ripple-valjer-bahnhof-nar-de-ska-etablera-sig-i-europa and this page also doesn't mention any public key. There is nothing at all on their website mentioning ANY validation key beyond that. So it has to be concluded that Bahnhof is not operating a validator or at least doesn't want to be associated with running one.

It would be trivial for Ripple to release also the signed statement(s) by the validator operators instead of just adding a tick mark on a website. They don't contain secrets and don't compromise any kind of security. Since this isn't done... well, Bahnhof certainly doesn't claim to operate any validator on their homepage and Ripple has some weird standards what a "domain" is (`ripple.com/build/xrp-test-net/` apparently is a "domain"?!), so their competence there is questionable at best.

> Right now, I can think of only two such identifiers

There's also Namecoin, ENS... even Ripple a few years back wanted to go into this direction with the whole `~name` thing. 

Certificates (e.g. in x509 format) and infrastructure around them already exist, getting one for a validation keypair shouldn't be too hard and Ripple could operate a CA that signs CSRs from validator operators fulfilling whatever their identification policy is.

There's also DNS TXT records if you want to go the domain route, XRPL has even a "domain" field for that purpose that could be used (at last).

> Right now, given a published list, the only way to associate a validator public key with an identifier is by visiting a website; this is a pain. Additionally, the data that website reports is not signed in an way.

https://xrpcharts.ripple.com/#/validators is signed (through TLS) by Ripple's server, but contains the convenient statement `Information is provided "as is" and solely for informational purposes only.` (another hint that their "Bahnhof" association is not to be trusted btw.). What's a bit more inconvenient is that the vl.ripple.com page contains raw public keys (in a base64 encoded, custom, undocumented format) and the "validator registry" lists them in base58.

> if the publisher of the list lies, there will be cryptographic proof of that lie and we can all decide what to do moving forward.

I'd claim that Ripple is currently lying about the www.bahnhof.se validator for the reasons listed above. It is impossible to prove a negative though and the positive proof (e.g. a signature by Bahnhof or a mention of their validation key on their website) is hard to get without cooperation by any party involved.

> It doesn't get us to the point of users being able to independently verify the domain by posting a signature or other cryptographic proof that binds the domain and the validator together, but that is something we are working on.

This is something that is already required for https://docs.google.com/forms/d/e/1FAIpQLScszfq7rRLAfArSZtvitCyl-VFA9cNcdnXLFjURsdCQ3gHW7w/viewform

> We are working on a way to allow people to independently verify the domain (by posting the signature) but in the mean time, I see this as a net win.

I see it as a premature optimization that is locking in expectations, introducing even more trust in validation list providers and narrowing down the design choices needlessly. There is little to no way to actually get a proof either way until you are releasing these signatures, so there's not much benefit in having a signed statement.

**Comment by nbougalis** [comment]
>Easy: None of them is, because Ba*h*nhof is not operating a validator.

You can have a serious conversation or you can joke about a typo.

>Proof: https://www.bahnhof.se/ripple.txt

That isn't proof of anything. Hosting a `ripple.txt` is **NOT** a prerequisite for operating a validator and never has been. It is true that an operator could **choose** to host such a file and could **choose** to list their validation public keys in that file, but again, it is **NOT** a prerequisite.

>Alternatively https://www.google.com/search?q=site%3Abahnhof.se+ripple also only finds https://www.bahnhof.se/press/press-releases/2017/07/18/fintechbolaget-ripple-valjer-bahnhof-nar-de-ska-etablera-sig-i-europa and this page also doesn't mention any public key. There is nothing at all on their website mentioning ANY validation key beyond that. So it has to be concluded that Bahnhof is not operating a validator or at least doesn't want to be associated with running one.

If that's your conclusion, that's fine. As it stands now, Ripple, as a validator list publisher, is **attesting** that they have verified that the validator `nHUFzgC9fDw2MEDaiv9JMdBFhtJ6DMKoUCpS8gPGi6tkfbqmTyis` is operated by Bahnhof. 

>It would be trivial for Ripple to release also the signed statement(s) by the validator operators instead of just adding a tick mark on a website. They don't contain secrets and don't compromise any kind of security. Since this isn't done... well, Bahnhof certainly doesn't claim to operate any validator on their homepage and Ripple has some weird standards what a "domain" is (ripple.com/build/xrp-test-net/ apparently is a "domain"?!), so their competence there is questionable at best.

First, re `ripple.com/build/xrp-test-net` being a "domain" I agree that the naming is not ideal; that is certainly **not** a domain, and it's something that should be fixed. But, honestly, if that's your concern you're grasping at straws and if you think that is an indicator of technical competency... well, I consider that an indicator of something as well. But I digress.

As I explained, we are working on making this information available, but I think this is an improvement over what is already there for two reasons:
1. It allows a validator list publisher to *cryptographically attest* to a domain name associated with a validator.
2. It allows someone querying the active lists of a `rippled` instance to get a human-readable and human-relevant name attached to each validator without having to visit a website or access some API that maps validator keys to identities.
To me, both of those are net gain.

>There's also Namecoin, ENS... even Ripple a few years back wanted to go into this direction with the whole ~name thing.

Re: ~name; Ripple is no longer going into that direction, and let's be honest here: if we tried to use that for validator names you'd be complaining even louder that you are complaining now.

As for Namecoin, ENS, etc, if you want to see that support, then code it up and submit a PR. Don't bash a PR by someone else.

>Certificates (e.g. in x509 format) and infrastructure around them already exist, getting one for a validation keypair shouldn't be too hard and Ripple could operate a CA that signs CSRs from validator operators fulfilling whatever their identification policy is.

Why don't **you** operate a CA and sign CSRs from validator operators? Then you can enforce whatever identification requirements you want.

>There's also DNS TXT records if you want to go the domain route, XRPL has even a "domain" field for that purpose that could be used (at last).

We've contemplated storing information in DNS TXT records, but there have been a number of issues associated with that, including the fact that getting buy-in to add records to DNS (or post a **ripple.txt**) isn't always easy, especially with larger organizations.

>I'd claim that Ripple is currently lying about the www.bahnhof.se validator for the reasons listed above. 

You know what? I'm done engaging with you.

**Comment by MarkusTeufelberger** [comment]
> That isn't proof of anything. Hosting a ripple.txt is NOT a prerequisite for operating a validator and never has been.

I agree, "proof" was maybe too strong of a wording. It would have been proof if they had published the file, its absence is not proving that they are or aren't running a validator.

> As it stands now, Ripple, as a validator list publisher, is attesting that they have verified that the validator nHUFzgC9fDw2MEDaiv9JMdBFhtJ6DMKoUCpS8gPGi6tkfbqmTyis is operated by Bahnhof.

On a page with `Information is provided "as is" and solely for informational purposes only.` in its footer and without publishing anything other than a `✓` character as proof. This PR would change that checkmark into a field in a signed json file, nothing more, nothing less.

> You know what? I'm done engaging with you.

:-(

I'm just trying to find out how this would work out:

> This change actually improves things: the publisher, whom the user trusts to publish a list, is attesting that a given validator corresponds to a particular domain and cryptographically signs that attestation; **if the publisher of the list lies, there will be cryptographic proof of that lie and we can all decide what to do moving forward.**

My reasoning is that it doesn't improve things, since there is no way on the outside to proof that someone is lying, only to claim that someone is lying, for example in the way I did, with the following reasoning/indications:

* Bahnhof doesn't list a validation pubkey on their site, in neither the canonical location (`ripple.txt`) or anywhere else (even though it would be simple/easy to do so, e.g. in the announcement blog post)
* The only "attestation" published by Ripple is a checkmark character on an `informational only` page
* That page also uses technical terms rather broadly, with the example that `ripple.com/build/xrp-test-net` is a "verified domain" as of today
* No signatures (that potentially already exist) are published by Ripple or Bahnhof, even though it would be easy and secure to do so
* Bahnhof is not shown on https://ripple.com/xrp/market-performance/ (but a lot of organizations that not even Ripple claims are validating are listed there, so maybe these logos represent something else or are just examples)

The indications for Bahnhof running a validator are:

* A blog post by Bahnhof from ~1 year ago that they are "part of a growing network of validating organizations"
* Ripple's claims (on the validator registry) that the validation pubkey that hashes + encodes to `nHUFzgC9fDw2MEDaiv9JMdBFhtJ6DMKoUCpS8gPGi6tkfbqmTyis` belongs to `www.bahnhof.se`

The latter is also the only mention of Bahnhof's public key.

You might (or rather: likely do) have more insight or trust in the process than me of course. Still, adding a domain name to the validation list blob doesn't make it any easier to prove or disprove if Bahnhof is validating with that pubkey or if Ripple is honest about this.

An "easy" proof that Ripple is lying would be a signed statement from the private key that they don't belong to Bahnhof (then again: It could have been stolen, which is outside of Ripple's control)
An easy proof that they are acting faithfully would be to attach the TLS cert, the validation pubkey signed with the cert key, the domain and the domain signed with the validation key to this blob. Then the trust would need to be placed with the TLS CA and some discrepancies (e.g. a wildcard cert being used to sign such a statement) could be found out.

Until then, while it might be possible to claim a lie, there is no proof of a lie (or of the absence of it!). This then boils down to trust, politics and people writing long posts on the internet, not cryptographic proof...

Tl;dr:
It is only useful for display purposes, it does not significantly increase the accountability but further increases the amount of trust necessary in the validation list provider.

**Comment by jnr101** [comment]
Thanks for this discussion and PR, it is imo necessary to create more openness in the decentralisation issue. For what it’s worth, I can understand both standpoints and I think that Markus is just taking the hypothetical standpoint of someone who is anti-Ripple. But besides that, nice discussion :-)


As for the domain attestation, I agree that a cryptographic Ripple-attestation does add value. 
Attestations come to play when linkage between real-world identities and ledger identities is necessary. These can in principle not be proven via the ledger. So you need 'some' party to attest for this information to be true. 

In case of using the domain name, certificate and published signatures, it is in the end the CA of the certificate that provides this attestation (provided the signatures are published somewhere/somehow by Ripple). It is a pretty good attestation, as the certificate is probably also linked to a company’s website. This adds value to the credibility of the attestation.

In case of only storing domain name in the validator list, it is Ripple who provides the attestation. (But also here a dependancy on some CA attesting that Ripple is Ripple :-)) And yes, of course it would be small step to include next to the domain name also the signatures. 

Point I want to make is, isn’t ‘domain’ field too specific? Maybe change it to something more generic, like ‘info’, ‘meta’ or ‘commitment’, so in the future you can also store other information in here.. Taking in to account that the validator list is primarily for the rippled servers and they will (can) not do anything with this info, you could e.g. use this field to include a hash of the proof where the proof is published somewhere else.

**Comment by nbougalis** [comment]
@MarkusTeufelberger: as I said I agree that the ideal system would be to provide cryptographic proof (i.e. the "statement" that a validator operator signs and the signature) allowing anyone to independently verify the signature. This is coming, and this PR wouldn't change or delay that, but it would add something which seems to me to be of value:
1. The ability for the publisher to _cryptographically attest_ to the domain name of an entry within the list itself.
2. The ability for anyone to archive and easily publish historical validation lists with the domain name included in the signed part of the data, without having to also store and index a webpage.
3. The ability for `rippled` to display the domain name—an identifier that is much more human friendly than a base58 validator public key.

Note that you were the person that asked us to publish historical signed lists, which we do at https://github.com/ripple/vl, by making the argument that having the signed information is important to ensure transparency and confidence. So I would have imagined you'd realize the value that including this extra information to the published lists add even if only for archiving purposes.

File: src/ripple/app/misc/impl/ValidatorList.cpp (hunk 2 above)
Comment (wilsonianb) [reply src/ripple/app/misc/impl/ValidatorList.cpp]: I'm assuming it's intentional, and I think it's the right thing to do. We're interested in a valid manifest even if it's on an invalid/malformed list.

**Comment by cryptobrad** [comment]
There seems to be no consensus about having a UNL publisher include (or attest to) a `domain` field. I think it's a good idea but @MarkusTeufelberger and @wilsonianb don't seem to agree.

What should next steps be here? Should I redo this PR without the domain field to clean up the awkward code or just ditch this altogether?

**Comment by nbougalis** [comment]
While I like the idea of a UNL publisher attesting to a domain name, I have something else in the works; I plan on publishing a little more about it in an issue here, to solicit feedback from the validator operator community as well as any interested third parties. I'll be sure to tag you on that.

So, I'd appreciate it if you could modify this PR to include your code improvements but without the support for the domain field. Thanks!

**Comment by cryptobrad** [comment]
Done. Thanks for getting back to me.

**Comment by scottschurr** [comment]
Here's a block of errors that Jenkins emitted on its most recent run:
```
In file included from /opt/jenkins/workspace/rippled-dev_PR-2619-S3CMQMLTSE3UHXS7CMAXE4XNBXQNMFS27FH56H4CSN4BY5DWZF2Q/src/ripple/unity/app_misc_impl.cpp:27:
/opt/jenkins/workspace/rippled-dev_PR-2619-S3CMQMLTSE3UHXS7CMAXE4XNBXQNMFS27FH56H4CSN4BY5DWZF2Q/src/ripple/app/misc/impl/ValidatorList.cpp:256:25: error: unknown type name 'ValidatorEntry'; did you mean 'ValidatorSite'?
            std::vector<ValidatorEntry> removed;
                        ^~~~~~~~~~~~~~
                        ValidatorSite
/opt/jenkins/workspace/rippled-dev_PR-2619-S3CMQMLTSE3UHXS7CMAXE4XNBXQNMFS27FH56H4CSN4BY5DWZF2Q/src/ripple/app/main/Application.h:70:7: note: 'ValidatorSite' declared here
class ValidatorSite;
      ^
In file included from /opt/jenkins/workspace/rippled-dev_PR-2619-S3CMQMLTSE3UHXS7CMAXE4XNBXQNMFS27FH56H4CSN4BY5DWZF2Q/src/ripple/unity/app_misc_impl.cpp:27:
/opt/jenkins/workspace/rippled-dev_PR-2619-S3CMQMLTSE3UHXS7CMAXE4XNBXQNMFS27FH56H4CSN4BY5DWZF2Q/src/ripple/app/misc/impl/ValidatorList.cpp:259:46: error: use of undeclared identifier 'comparator'
                std::back_inserter(removed), comparator);
                                             ^
/opt/jenkins/workspace/rippled-dev_PR-2619-S3CMQMLTSE3UHXS7CMAXE4XNBXQNMFS27FH56H4CSN4BY5DWZF2Q/src/ripple/app/misc/impl/ValidatorList.cpp:261:31: error: member access into incomplete type 'const ripple::ValidatorSite'
                keyListings_[r.pubkey]--;
                              ^
/opt/jenkins/workspace/rippled-dev_PR-2619-S3CMQMLTSE3UHXS7CMAXE4XNBXQNMFS27FH56H4CSN4BY5DWZF2Q/src/ripple/app/main/Application.h:70:7: note: forward declaration of 'ripple::ValidatorSite'
class ValidatorSite;
      ^
In file included from /opt/jenkins/workspace/rippled-dev_PR-2619-S3CMQMLTSE3UHXS7CMAXE4XNBXQNMFS27FH56H4CSN4BY5DWZF2Q/src/ripple/unity/app_misc_impl.cpp:27:
/opt/jenkins/workspace/rippled-dev_PR-2619-S3CMQMLTSE3UHXS7CMAXE4XNBXQNMFS27FH56H4CSN4BY5DWZF2Q/src/ripple/app/misc/impl/ValidatorList.cpp:261:32: error: expected ']'
                keyListings_[r.pubkey]--;
                               ^
/opt/jenkins/workspace/rippled-dev_PR-2619-S3CMQMLTSE3UHXS7CMAXE4XNBXQNMFS27FH56H4CSN4BY5DWZF2Q/src/ripple/app/misc/impl/ValidatorList.cpp:261:29: note: to match this '['
                keyListings_[r.pubkey]--;
                            ^
/opt/jenkins/workspace/rippled-dev_PR-2619-S3CMQMLTSE3UHXS7CMAXE4XNBXQNMFS27FH56H4CSN4BY5DWZF2Q/src/ripple/app/misc/impl/ValidatorList.cpp:266:25: error: unknown type name 'ValidatorEntry'; did you mean 'ValidatorSite'?
            std::vector<ValidatorEntry> added;
                        ^~~~~~~~~~~~~~
                        ValidatorSite
/opt/jenkins/workspace/rippled-dev_PR-2619-S3CMQMLTSE3UHXS7CMAXE4XNBXQNMFS27FH56H4CSN4BY5DWZF2Q/src/ripple/app/main/Application.h:70:7: note: 'ValidatorSite' declared here
class ValidatorSite;
      ^
In file included from /opt/jenkins/workspace/rippled-dev_PR-2619-S3CMQMLTSE3UHXS7CMAXE4XNBXQNMFS27FH56H4CSN4BY5DWZF2Q/src/ripple/unity/app_misc_impl.cpp:27:
/opt/jenkins/workspace/rippled-dev_PR-2619-S3CMQMLTSE3UHXS7CMAXE4XNBXQNMFS27FH56H4CSN4BY5DWZF2Q/src/ripple/app/misc/impl/ValidatorList.cpp:270:31: error: member access into incomplete type 'const ripple::ValidatorSite'
                keyListings_[a.pubkey]++;
                              ^
/opt/jenkins/workspace/rippled-dev_PR-2619-S3CMQMLTSE3UHXS7CMAXE4XNBXQNMFS27FH56H4CSN4BY5DWZF2Q/src/ripple/app/main/Application.h:70:7: note: forward declaration of 'ripple::ValidatorSite'
class ValidatorSite;
      ^
In file included from /opt/jenkins/workspace/rippled-dev_PR-2619-S3CMQMLTSE3UHXS7CMAXE4XNBXQNMFS27FH56H4CSN4BY5DWZF2Q/src/ripple/unity/app_misc_impl.cpp:27:
/opt/jenkins/workspace/rippled-dev_PR-2619-S3CMQMLTSE3UHXS7CMAXE4XNBXQNMFS27FH56H4CSN4BY5DWZF2Q/src/ripple/app/misc/impl/ValidatorList.cpp:270:32: error: expected ']'
                keyListings_[a.pubkey]++;
                               ^
/opt/jenkins/workspace/rippled-dev_PR-2619-S3CMQMLTSE3UHXS7CMAXE4XNBXQNMFS27FH56H4CSN4BY5DWZF2Q/src/ripple/app/misc/impl/ValidatorList.cpp:270:29: note: to match this '['
                keyListings_[a.pubkey]++;
```

**Comment by nbougalis** [comment]
Actually, @cryptobrad, I was working on something else that might benefit from having the `domain` field in the manifest itself.

Would you object to "readding" it, by restoring your previous commit? Thanks.
